docs/decisions/0033-rag-in-sk.md

```diff
@@ -0,0 +1,297 @@
+    .Build();
+
+kernel.ImportPluginFromObject(new TextMemoryPlugin(memory));
+
+var result = await kernel.InvokePromptAsync("{{recall 'Company budget by year'}} What is my budget for 2024?");
+```
+
+Custom user plugin:
+
+```csharp
+public class MyDataPlugin
+{
+    [KernelFunction("search")]
+    public async Task<string> SearchAsync(string query)
+    {
+        // Make a call to vector DB and return results.
+        // Here developer can use already existing .NET SDK from specific vector DB provider.
+        // It's also possible to re-use Semantic Kernel memory connector directly here: 
+        // new ChromaMemoryStore(...).GetNearestMatchAsync(...)
+    }
+}
+
+var kernel = Kernel.CreateBuilder()
+    .AddOpenAIChatCompletion("model-id", "api-key")
+    .Build();
+
+kernel.ImportPluginFromType<MyDataPlugin>();
+
+var result = await kernel.InvokePromptAsync("{{search 'Company budget by year'}} What is my budget for 2024?");
+```
+
+The reason why custom user plugin is more flexible than `TextMemoryPlugin` is because `TextMemoryPlugin` requires all vector DBs to implement `IMemoryStore` interface with disadvantages described above, while custom user plugin can be implemented in a way of developer's choice. There won't be any restrictions on DB record schema or requirement to implement specific interface.
+
+### Option 3 [Partially supported] - Prompt concatenation using Prompt Filter
+
+This option is similar to Option 1, but prompt concatenation will happen on Prompt Filter level:
+
+Prompt filter:
+
+```csharp
+public sealed class MyPromptFilter : IPromptFilter
+{
+    public void OnPromptRendering(PromptRenderingContext context)
+    {
+        // Handling of prompt rendering event...
+    }
+
+    public void OnPromptRendered(PromptRenderedContext context)
+    {
+        var data = "some data";
+        var builder = new StringBuilder();
+
+        builder.AppendLine(data);
+        builder.AppendLine(context.RenderedPrompt);
+
+        // Override rendered prompt before sending it to AI and include data
+        context.RenderedPrompt = builder.ToString();
+    }
+}
+```
+
+Usage:
+
+```csharp
+var kernel = Kernel.CreateBuilder()
+    .AddOpenAIChatCompletion("model-id", "api-key")
+    .Build();
+
+kernel.PromptFilters.Add(new MyPromptFilter());
+
+var result = await kernel.InvokePromptAsync("What is my budget for 2024?");
+```
+
+From the usage perspective, prompt will contain just user query without additional data. The data will be added to the prompt behind the scenes.
+
+The reason why this approach is **partially supported** is because a call to vector DB most probably will be an asynchronous, but current Kernel filters don't support asynchronous scenarios. So, in order to support asynchronous calls, new type of filters should be added to Kernel: `IAsyncFunctionFilter` and `IAsyncPromptFilter`. They will be the same as current `IFunctionFilter` and `IPromptFilter` but with async methods.
```

Comment:
Good point, that sounds like something that may need improvement indeed. Manipulating/filtering the prompt based on information/decision-making that needs to be accessed asynchronously sounds like an important capability.

Reply:
Yes, I think we will add support for asynchronous filters independently from memory-related work, since it's important feature that should cover asynchronous scenarios for filtering.